pinot-core/src/main/java/org/apache/pinot/core/plan/FilterPlanNode.java

```diff
@@ -52,19 +52,30 @@
 import org.apache.pinot.segment.spi.index.reader.JsonIndexReader;
 import org.apache.pinot.segment.spi.index.reader.NullValueVectorReader;
 import org.apache.pinot.spi.exception.BadQueryRequestException;
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
 
 
 public class FilterPlanNode implements PlanNode {
   private final IndexSegment _indexSegment;
   private final QueryContext _queryContext;
   private final int _numDocs;
+  private final MutableRoaringBitmap _validDocIds;
 
   // Cache the predicate evaluators
   private final Map<Predicate, PredicateEvaluator> _predicateEvaluatorMap = new HashMap<>();
 
   public FilterPlanNode(IndexSegment indexSegment, QueryContext queryContext) {
     _indexSegment = indexSegment;
     _queryContext = queryContext;
+
+    // Snapshot validDocIds *before* _numDocs since otherwise newest upserts could get lost
```

Comment:
Taking snapshot when creating the plan nodes can potentially cause performance issue since it is done single-threaded

Reply:
Good point. From what I could see the FilterPlanNode constructor call site is currently always adjacent to the run method invocation (DocSetIdPlanNode#run, AggregationPlanNode#run, AggregationGroupByPlanNode#run, AggregationGroupByOrderByPlanNode#run) so I don't think that it could currently make a difference whether snapshotting happens in constructor or run. However, the intent of other operators seems definitely to be that run happens during parallelized execution. From that perspective the approach in #8018 to only snapshot in the run method is definitely preferable.

Only question I had was what the original reason was behind the checkpointing in the constructor (given the same argument as above it seems to make no practical difference at the moment :)?

Reply:
Oh you are right. We recently changed the logic of `FilterPlanNode`, and it is not constructed in the plan maker, so there is no practical difference now

Reply:
In any case I know how to make snapshots faster in the library given that the bitmap is only ever appended to, so we could do that work if it ever showed up in profiles.

Reply:
@richardstartin It is not only appended to, but also modified. We can probably keep the writer in the `ThreadSafeMutableRoaringBitmap` and only create the bitmap during the snapshot if that gives better performance